File: java/src/main/java/net/imagini/aim/partition/AimPartitionServer.java
package net.imagini.aim.partition;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

import net.imagini.aim.Pipe;


public class AimPartitionServer extends Thread {
    ServerSocket controllerListener;
    private Thread controllerAcceptor;
    private AimPartition partition;
    public final int port;

    public AimPartitionServer(AimPartition partition, int port) throws IOException {
        this.partition = partition;
        this.port = port;

        controllerListener = new ServerSocket(port);
        controllerAcceptor = new Thread() {
            @Override public void run() {
                System.out.println("Aim Partition Server accepting connections on port " + AimPartitionServer.this.port + " " + AimPartitionServer.this.partition.schema);
                    while(true) {
                        Socket socket;
                        try {
                            socket = controllerListener.accept();
                            if (interrupted()) break;
                        } catch (IOException e) {
                            System.out.println(e.getMessage());
                            break;
                        }
                        Pipe pipe;
                        try {
                            if (interrupted()) break;
                            pipe = Pipe.open(socket);
                            System.out.println(pipe.protocol + " connection from " + socket.getRemoteSocketAddress().toString());
                            switch(pipe.protocol) {
                                //TODO case BINARY: new ReaderThread(); break;
                                case LOADER: new AimPartitionServerLoaderSession(AimPartitionServer.this.partition,pipe).start(); break;
                                case QUERY: new AimPartitionServerQuerySession(AimPartitionServer.this.partition,pipe).start(); break;
                                default: System.out.println("Unsupported protocol request " + pipe.protocol);
                            }
                        } catch (IOException e) {
                            System.out.println("Aim Server at port "+AimPartitionServer.this.port +" failed to establish client connection: " + e.getMessage());
                            continue;
                        }
                    }
            }
        };
    }

    @Override public void run() {
        controllerAcceptor.start();
        try {
            controllerAcceptor.join();
        } catch (InterruptedException e1) {
            controllerAcceptor.interrupt();
        } finally {
            close();
            System.out.println("Aim Server at port "+AimPartitionServer.this.port + " shut down.");
        }
    }

    public void close() {
        try {
            controllerListener.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        //TODO close all connections
    }
}

File: java/src/main/java/net/imagini/aim/partition/AimPartitionServer.scala
package net.imagini.aim.partition

import net.imagini.aim.types.AimSchema
import net.imagini.aim.types.SortOrder
import java.net.ServerSocket
import java.io.IOException
import net.imagini.aim.tools.Protocol
import net.imagini.aim.tools.Pipe

object AimPartitionServer extends App {

  var port = 4000
  var filename: String = null
  var separator = "\n"
  var gzip = false
  var schema: Option[AimSchema] = None
  val argsIterator = args.iterator
  var limit = 0L
  var segmentSize = 10485760
  while (argsIterator.hasNext) {
    argsIterator.next match {
      case "--segment-size" ⇒ segmentSize = argsIterator.next.toInt
      case "--port"         ⇒ port = argsIterator.next.toInt
      case "--separator"    ⇒ separator = argsIterator.next
      case "--gzip"         ⇒ gzip = true;
      case "--schema"       ⇒ schema = Some(AimSchema.fromString(argsIterator.next))
      case "--filename"     ⇒ filename = argsIterator.next
      case arg: String      ⇒ println("Unknown argument " + arg)
    }
  }

  val key = schema.get.name(0)
  val sortOrder = SortOrder.ASC
  val partition = new AimPartition(schema.get, segmentSize, key, sortOrder)
  val server = new AimPartitionServer(partition, port)

  new AimPartitionLoader("localhost", port, schema.get, separator, filename, gzip).processInput

}

class AimPartitionServer(val partition: AimPartition, val port: Int) {

  val controllerListener = new ServerSocket(port)
  val controllerAcceptor = new Thread {
    override def run = {
      println("Aim Partition Server accepting connections on port " + port + " " + partition.schema)
      try {
        while (!isInterrupted) {
          val socket = controllerListener.accept
          try {
            val pipe = Pipe.open(socket)
            println("Aim Partition Server " + pipe.protocol + " connection from " + socket.getRemoteSocketAddress.toString)
            pipe.protocol match {
              //TODO case Protocol.BINARY: new ReaderThread(); break;
              //TODO case Protocol.MAPREDUCE: new ReaderThread(); break;
              case Protocol.LOADER ⇒ new AimPartitionServerLoaderSession(partition, pipe).start
              case Protocol.QUERY  ⇒ new AimPartitionServerQuerySession(partition, pipe).start
              case _               ⇒ println("Unsupported protocol request " + pipe.protocol)
            }
          } catch {
            case e: IOException ⇒ {
              println("Aim Partition at port " + port + " failed to establish client connection: " + e.getMessage)
            }
          }
        }
      } catch {
        case e: Throwable ⇒ {
          println(e.getMessage)
          controllerListener.close
        }
      }
    }
  }

  controllerAcceptor.start

  def close = {
    controllerAcceptor.interrupt
    controllerListener.close
    System.out.println("Aim Server at port "+port + " shut down.")
  }

} 